src/utils/parse.rs

```diff
@@ -83,3 +84,286 @@ where
+            (ParseState::Index, b':') => {
+                if token.is_empty() {
+                    return Err(ParseVectorError::TooShortNumber { position });
+                }
+                let s = unsafe { std::str::from_utf8_unchecked(&token[1..]) };
+                let index = s
+                    .parse::<u32>()
+                    .map_err(|_| ParseVectorError::BadParsing { position })?;
+                indexes.push(index);
+                token.clear();
+                state = ParseState::Value;
+            }
+            (ParseState::Value, b'0'..=b'9' | b'a'..=b'z' | b'A'..=b'Z' | b'.' | b'+' | b'-') => {
+                if token.is_empty() {
+                    token.push(b'$');
+                }
+                if token.try_push(c).is_err() {
+                    return Err(ParseVectorError::TooLongNumber { position });
+                }
+            }
+            (ParseState::Value, b',') => {
+                if token.is_empty() {
+                    return Err(ParseVectorError::TooShortNumber { position });
+                }
+                let s = unsafe { std::str::from_utf8_unchecked(&token[1..]) };
+                let num = f(s).ok_or(ParseVectorError::BadParsing { position })?;
+                values.push(num);
+                token.clear();
+                state = ParseState::Comma;
+            }
+            (ParseState::Value, b'}') => {
+                if token.is_empty() {
+                    return Err(ParseVectorError::TooShortNumber { position });
+                }
+                let s = unsafe { std::str::from_utf8_unchecked(&token[1..]) };
+                let num = f(s).ok_or(ParseVectorError::BadParsing { position })?;
+                values.push(num);
+                token.clear();
+                state = ParseState::Splitter;
+            }
+            (ParseState::Splitter, b'/') => {
+                state = ParseState::Length;
+            }
+            (ParseState::Length, b'0'..=b'9') => {
+                if token.is_empty() {
+                    token.push(b'$');
+                }
+                if token.try_push(c).is_err() {
+                    return Err(ParseVectorError::TooLongNumber { position });
+                }
+            }
+            (_, _) => {
+                return Err(ParseVectorError::BadCharacter { position });
+            }
+        }
+    }
+    if state != ParseState::Length {
+        return Err(ParseVectorError::BadParsing {
+            position: input.len(),
+        });
+    }
+    if token.is_empty() {
+        return Err(ParseVectorError::TooShortNumber {
+            position: input.len(),
+        });
+    }
+    let s = unsafe { std::str::from_utf8_unchecked(&token[1..]) };
+    let dims = s
+        .parse::<usize>()
+        .map_err(|_| ParseVectorError::BadParsing {
+            position: input.len(),
+        })?;
+
+    let mut indices = (0..indexes.len()).collect::<Vec<_>>();
+    indices.sort_by_key(|&i| &indexes[i]);
```

Comment:
I feel bad but there is surely no a function called `sort_with_indices` in std. But why call `sort` twice? A permutation can be decomposed into one or more disjoint cycles and we could use a loop to get all indexes and values in right position.